src/core/solvers/ipopt/ipopt-iface.cpp

```diff
@@ -0,0 +1,665 @@
+        for (std::size_t idx_col = 0; idx_col < ndxi; idx_col++) {
+          values[idx_value] = datas_[0]->Jg_ic(idx_row, idx_col);
+          idx_value++;
+        }
+      }
+    }
+  }
+
+  return true;
+}
+
+bool IpoptInterface::eval_h(Ipopt::Index n, const Ipopt::Number *x, bool new_x, Ipopt::Number obj_factor,
+                            Ipopt::Index m, const Ipopt::Number *lambda, bool new_lambda, Ipopt::Index nele_hess,
+                            Ipopt::Index *iRow, Ipopt::Index *jCol, Ipopt::Number *values) {
+  assert_pretty(n == nvar_, "Inconsistent number of decision variables");
+  assert_pretty(m == nconst_, "Inconsistent number of constraints");
+
+  if (values == NULL) {
+    // return the structure. This is a symmetric matrix, fill the lower left
+    // triangle only
+
+    // Running Costs
+    const std::vector<boost::shared_ptr<ActionModelAbstract> > &models = problem_->get_runningModels();
+    std::size_t idx_value = 0;
+    std::size_t ixu = 0;
+    for (std::size_t t = 0; t < T_; t++) {
+      const std::size_t ndxi = models[t]->get_state()->get_ndx();
+      const std::size_t nui = models[t]->get_nu();
+      for (std::size_t idx_row = 0; idx_row < ndxi + nui; idx_row++) {
+        for (std::size_t idx_col = 0; idx_col < ndxi + nui; idx_col++) {
+          // We need the lower triangular matrix
+          if (idx_col > idx_row) {
+            break;
+          }
+          iRow[idx_value] = ixu + idx_row;
+          jCol[idx_value] = ixu + idx_col;
+          idx_value++;
+        }
+      }
+      ixu += ndxi + nui;
+    }
+
+    {
+      // Terminal costs
+      const std::size_t ndxi = problem_->get_terminalModel()->get_state()->get_ndx();
+      for (std::size_t idx_row = 0; idx_row < ndxi; idx_row++) {
+        for (std::size_t idx_col = 0; idx_col < ndxi; idx_col++) {
+          // We need the lower triangular matrix
+          if (idx_col > idx_row) {
+            break;
+          }
+          iRow[idx_value] = ixu + idx_row;
+          jCol[idx_value] = ixu + idx_col;
+          idx_value++;
+        }
+      }
+    }
+
+    assert_pretty(idx_value == nele_hess,
+                  "Number of Hessian elements set does not coincide with the total non-zero Hessian values");
+  } else {
+    // return the values. This is a symmetric matrix, fill the lower left
+    // triangle only
+    std::size_t idx_value = 0;
+    std::size_t ixu = 0;
+    // Running Costs
+    for (std::size_t t = 0; t < T_; t++) {
+      const boost::shared_ptr<ActionModelAbstract> &model = problem_->get_runningModels()[t];
+      const boost::shared_ptr<ActionDataAbstract> &data = problem_->get_runningDatas()[t];
+      const std::size_t ndxi = model->get_state()->get_ndx();
+      const std::size_t nui = model->get_nu();
+
+      datas_[t]->dx = Eigen::VectorXd::Map(x + ixu, ndxi);
+      datas_[t]->u = Eigen::VectorXd::Map(x + ixu + ndxi, nui);
+
+      model->get_state()->integrate(xs_[t], datas_[t]->dx, datas_[t]->x);
```

Comment:
We shouldn't need to perform this integration again, right?
This should be done in the function that compute the cost or constraints.

Most likely this comment can be applied in other places.

Reply:
Same as other optimization related comments. Without knowing what is called first we cannot do this kind of optimizations.